src/controllers/midi/midisourceclock.cpp

```diff
@@ -0,0 +1,117 @@
+    m_bRunning = true;
+    m_iFilled = 0;
+    m_iRingBufferPos = 0;
+}
+
+void MidiSourceClock::stop() {
+    m_bRunning = false;
+}
+
+void MidiSourceClock::tick() {
+    // Update the ring buffer and calculate new bpm.  Update the last beat time
+    // if we are on a beat.
+
+    if (!m_bRunning) {
+        qDebug() << "MidiSourceClock: Got clock tick but not started, starting now.";
+        start();
+    }
+
+    // Ringbuffer filling.
+    // TODO(owen): We should have a ringbuffer convenience class.
+    const qint64 lastTickTime = m_pClock->now();
+    m_iTickRingBuffer[m_iRingBufferPos] = lastTickTime;
+    m_iRingBufferPos = (m_iRingBufferPos + 1) % kRingBufferSize;
+    if (m_iFilled < kRingBufferSize) {
+        ++m_iFilled;
+    }
+
+    // If this tick is a beat mark, record it, even if we have very few samples.
+    if (m_iRingBufferPos % kPulsesPerQuarter == 0) {
+        m_iLastBeatTime = lastTickTime;
+    }
+
+    // Figure out the bpm if we have enough samples.
+    if (m_iFilled > 2) {
+        qint64 earlyTickTime = 0;
+        if (m_iFilled < kRingBufferSize) {
+            earlyTickTime = m_iTickRingBuffer[0];
+        } else {
+            // In a filled ring buffer, the earliest tick is the next one that
+            // will get filled.
+            earlyTickTime = m_iTickRingBuffer[m_iRingBufferPos];
+        }
+        m_dBpm = calcBpm(earlyTickTime, lastTickTime, m_iFilled);
+    }
+}
+
+// static
+double MidiSourceClock::calcBpm(qint64 early_tick, qint64 late_tick,
+                                int tick_count) {
+    // Get the elapsed time between the latest tick and the earliest tick
+    // and divide by the number of ticks in the buffer to get bpm.
+
+    // If we have too few samples, we can't calculate a bpm, so return 0.0.
+    DEBUG_ASSERT_AND_HANDLE(tick_count >= 2) {
+        qWarning() << "MidiSourceClock::calcBpm called with too few ticks";
+        return 0.0;
+    }
+
+    DEBUG_ASSERT_AND_HANDLE(late_tick >= early_tick) {
+        qWarning() << "MidiSourceClock asked to calculate beat percentage but "
+                   << "late_tick < early_tick:" << late_tick << early_tick;
+        return 0.0;
+    }
+
+    const double elapsed_mins = static_cast<double>(late_tick - early_tick)
+            / (60.0 * 1e9);
+
+    // We subtract one since two time values denote a single span of time --
+    // so a filled value of 3 indicates 2 tick periods, etc.
+    const double bpm = static_cast<double>(tick_count - 1) / kPulsesPerQuarter
+                       / elapsed_mins;
+
+    return math_clamp(bpm, kMinMidiBpm, kMaxMidiBpm);
+}
+
+// static
+double MidiSourceClock::beatPercentage(const qint64 last_beat, const qint64 now,
+                                       const double bpm) {
+    DEBUG_ASSERT_AND_HANDLE(now >= last_beat) {
+        qWarning() << "MidiSourceClock asked to calculate beat percentage but "
+                   << "now < last_beat:" << now << last_beat;
+        return 0.0;
```

Comment:
we can make a guess, but by the API I've written this situation is invalid and indicates a bad call upstream.